### ISSUE TITLE
Parameters for LSTM model on Reddit dataset

### ISSUE THREAD

**Issue by bkj** [williamleif/GraphSAGE#9]
Do you happen to have the parameters for the supervised LSTM model on the Reddit dataset that yields 0.954? I'm running 

```
python -m graphsage.supervised_train \
    --train_prefix ./data/reddit \
    --model graphsage_seq \
    --sigmoid
```

on my machine, and getting results closer to  ~0.92.  

Similar question, do you have any sense of the statistical significance in the differences between the various supervised GraphSAGE models? `mean`, `LSTM` and `pool` appear to outperform `gcn`, and certainly outperform the baseline models, but is there strong evidence that `LSTM` performs better than `mean`, given different random starts, etc?  (I ask because `LSTM` is significantly slower, so it'd be nice if I didn't have to use it...)

Thanks

**Comment by williamleif** [comment]
I don't remember the optimal parameters of the top of my head, and I'll have to spend a bit of time looking through the logs on our server to get them, but one thing I can say is that you shouldn't need the "sigmoid" flag for the Reddit data, and this might be giving you a slight performance hit. 

Re statistical significance: GCN is definitely significantly worse than the others, but between the other methods the results are more equivocal. IIRC the Reddit data is also the most high-variance in terms of performance results (whereas the citation data was the most stable). I can say that in (not-yet-public) follow-up work (industry collaboration with lots of compute power), we found that a pooling aggregator was definitively the best. 

So yeah, my recommendation is that it is fine to not use the LSTM model. It did surprisingly well, but it is not *definitely* not significantly better than the pooling model and not worth the extra compute.

We hope to add more discussion on this in the camera-ready. (And hopefully also some more insight into the variance of the results.. though compute time is always an issue haha).

**Comment by williamleif** [comment]
btw it looks like parameter settings were 0.001 for learning rate and using the "small" version of the model. (i.e., --model_size "small", which is the default). I also just ran it twice and got 0.954 and 0.955. 

So you need to change the learning rate and remove the --sigmoid flag.

Let me know if you still have troubles! 

Also, I should mention, as a follow-up to the above, the pooling aggregator does strangely poorly on the Reddit data --- much worse than any other dataset we've tried --- so if you are particularly interested in that benchmark, I'd recommend just using the mean aggregator.

**Comment by bkj** [comment]
Ah great -- thanks for the quick response.  The `sigmoid` flag is probably the issue -- I thought it was a multiclass problem, for some reason.

FYI, if you're interested, I did a bunch of refactoring and porting in this fork:  https://github.com/bkj/pytorch-graphsage

Results are pretty consistent w/ the original code on the Reddit dataset -- have yet to test the other benchmarks.

Have only looked at the supervised methods so far, but the `master` branch is a port to Pytorch, and the `tf-simple` branch is a stripped down Tensorflow version.  There may be opportunity to contribute some stuff from here back to this repo in the future.  Also implemented an attention-based aggregator, which is maybe interesting (seems to work pretty well in the limited tests)

**Comment by williamleif** [comment]
That's awesome! The refactoring/cleaning is nice and I am happy to see a pytorch fork. We are doing a lot of new development in pytorch, and those projects will hopefully be public soon! It is really just a lot easier to use dynamic compute graphs in this setting. (Though the runtime is bit slower with pytorch in my experience). 

It would be great if you want to contribute back to master; definitely make a pull request if you think a new feature or code change would be helpful. We are very happy to have community collaboration on this. And, in fact, an attention-based aggregator is something we've been meaning to add. 

(Anyways, I am going to close this issue, but I am happy to chat about particular extensions, design choices etc if you want, either via email, pull requests, or feature-request issues).

**Comment by bkj** [comment]
Cool, sounds good.  I'm in a period of making a bunch of changes -- I'll try to contribute back once things stabilize on my end.  I'll also likely reach out about extension, etc, once I have a chance to try out a couple other things I've been wanting to do.
